File: app/src/main/java/com/centaury/cataloguemovie/ui/movie/MovieViewModel.java
package com.centaury.cataloguemovie.ui.movie;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;

import com.centaury.cataloguemovie.data.CatalogueRepository;
import com.centaury.cataloguemovie.data.datasource.MovieDataSource;
import com.centaury.cataloguemovie.data.datasource.MovieDataSourceFactory;

import java.util.List;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import javax.inject.Inject;

/**
 * Created by Centaury on 10/7/2019.
 */
public class MovieViewModel extends ViewModel {

    private CatalogueRepository catalogueRepository;
    private LiveData<PagedList<MovieResultsItem>> movies;
    private LiveData<Boolean> loadingState;
    private LiveData<Boolean> loadMoreLoadingState;

    @Inject
    public MovieViewModel(CatalogueRepository catalogueRepository) {
        this.catalogueRepository = catalogueRepository;
        init();
    }

    private void init() {
        Executor executor = Executors.newFixedThreadPool(5);
        MovieDataSource movieDataSource = new MovieDataSource();
        MovieDataSourceFactory movieDataSourceFactory = new MovieDataSourceFactory(movieDataSource);

        PagedList.Config pagedListConfig =
                (new PagedList.Config.Builder())
                        .setEnablePlaceholders(true)
                        .setInitialLoadSizeHint(1)
                        .setPrefetchDistance(10)
                        .setPageSize(10).build();

        movies = (new LivePagedListBuilder<>(movieDataSourceFactory, pagedListConfig))
                .setFetchExecutor(executor)
                .build();

        loadingState = movieDataSource.getLoadingState();
        loadMoreLoadingState = movieDataSource.getLoadMoreLoadingState();
    }

    LiveData<PagedList<MovieResultsItem>> getMovies() {
        return movies;
    }

    LiveData<List<GenresItem>> getGenreMovie(String language) {
        return catalogueRepository.getGenreMovie(language);
    }

    LiveData<Boolean> getLoadingState() {
        return loadingState;
    }

    LiveData<Boolean> getLoadMoreLoadingState() {
        return loadMoreLoadingState;
    }
}

File: app/src/main/kotlin/com/centaury/cataloguemovie/ui/movie/MovieViewModel.kt
package com.centaury.cataloguemovie.ui.movie

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.centaury.cataloguemovie.utils.LoaderState
import com.centaury.domain.UseCase
import com.centaury.domain.model.Genre
import com.centaury.domain.model.Movie
import com.centaury.domain.movies.interactor.GetDiscoveryMovie
import com.centaury.domain.movies.interactor.GetGenreMovie
import javax.inject.Inject

/**
 * Created by Centaury on 10/7/2019.
 */
class MovieViewModel @Inject constructor(
    private val getDiscoveryMovie: GetDiscoveryMovie,
    private val getGenreMovie: GetGenreMovie
) : ViewModel(), MovieContract {

    private val _state = MutableLiveData<LoaderState>()
    val state: LiveData<LoaderState>
        get() = _state

    private val _result = MutableLiveData<List<Movie>>()
    val result: LiveData<List<Movie>>
        get() = _result

    private val _error = MutableLiveData<String>()
    val error: LiveData<String>
        get() = _error

    private val _resultGenre = MutableLiveData<List<Genre>>()
    val resultGenre: LiveData<List<Genre>>
        get() = _resultGenre

    private val _errorGenre = MutableLiveData<String>()
    val errorGenre: LiveData<String>
        get() = _errorGenre

    init {
        getDiscoverMovieContract()
        getGenreMovieContract()
    }

    override fun getDiscoverMovieContract() {
        _state.value = LoaderState.ShowLoading
        getDiscoveryMovie.execute(UseCase.None(), onSuccess = {
            _state.value = LoaderState.HideLoading
            _result.postValue(it)
        }, onError = {
            _state.value = LoaderState.HideLoading
            _error.postValue(it.message)
        })
    }

    override fun getGenreMovieContract() {
        _state.value = LoaderState.ShowLoading
        getGenreMovie.execute(UseCase.None(), onSuccess = {
            _state.value = LoaderState.HideLoading
            _resultGenre.postValue(it)
        }, onError = {
            _state.value = LoaderState.HideLoading
            _errorGenre.postValue(it.message)
        })
    }

    override fun onCleared() {
        super.onCleared()
        getDiscoveryMovie.dispose()
        getGenreMovie.dispose()
    }
}